File: csvdiff/engine.py
"""Composite-key CSV comparison.

Engine: DuckDB. It streams both files from disk, infers delimiters, hash-joins
on the composite key and spills to disk when data exceeds RAM, so multi-GB files
are fine. It is a required dependency rather than one of several: the byte-level
ports in cpp/, rust/, zig/, java/ and go/ are where the alternatives live, and
they are held to this same contract.

Result contract (JSON-serialisable dict) — consumed by report.py and by any
other sink (JSON sidecar, email, CI gate):

{
  "meta":    {a, b, key, compared, only_in_a, only_in_b, options, engine, seconds, generated}
  "counts":  {a_rows, b_rows, a_keys, b_keys, matched, unchanged, changed, added, removed,
              a_dup_keys, a_dup_rows, b_dup_keys, b_dup_rows}
  "columns": [{name, changed, blanked, filled}]           # per compared column
  "changed": {"cols": key, "rows": [[k1..kn, [[colIdx, old, new], ...]], ...], "truncated": bool}
  "added":   {"cols": key + compared, "rows": [[...]], "truncated": bool}
  "removed": same as added
  "dup_a":   {"cols": key + ["count"], "rows": [[...]], "truncated": bool}
  "dup_b":   same
}
"""
from __future__ import annotations

import os
import time
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from typing import Any


@dataclass
class Options:
    key: list[str]
    compare: list[str] | None = None      # None -> every common non-key column
    ignore: list[str] = field(default_factory=list)
    trim: bool = False                    # strip surrounding whitespace before comparing
    ignore_case: bool = False
    empty_is_null: bool = False           # treat '' and NULL as equal
    tolerance: float = 0.0                # numeric tolerance (abs) when both sides parse as numbers
    max_rows: int = 50_000                # rows embedded per section in the report
    delimiter: str | None = None          # None -> auto-detect
    encoding: str = "utf-8"
    engine: str = "auto"                  # auto | duckdb
    threads: int | None = None
    memory_limit: str | None = None       # e.g. "4GB" (DuckDB only)
    export_dir: str | None = None         # write full, uncapped changed/added/removed CSVs here


class CompareError(ValueError):
    pass


def compare(a_path: str, b_path: str, opt: Options) -> dict[str, Any]:
    if not opt.key:
        raise CompareError("At least one key column is required.")
    for p in (a_path, b_path):
        if not os.path.isfile(p):
            raise CompareError(f"File not found: {p}")

    t0 = time.perf_counter()
    engine = "duckdb" if opt.engine == "auto" else opt.engine
    if engine != "duckdb":
        raise CompareError(f"Unknown engine: {engine}")
    try:
        import duckdb  # noqa: F401
    except ImportError as exc:
        raise CompareError(
            "DuckDB is required: pip install duckdb. It is the only engine this "
            "implementation carries; the byte-level ports are in cpp/, rust/, zig/, "
            "java/ and go/."
        ) from exc
    result = _compare_duckdb(a_path, b_path, opt)

    result["meta"].update({
        "a": {"name": os.path.basename(a_path), "path": os.path.abspath(a_path),
              "size": os.path.getsize(a_path)},
        "b": {"name": os.path.basename(b_path), "path": os.path.abspath(b_path),
              "size": os.path.getsize(b_path)},
        "engine": engine,
        "seconds": round(time.perf_counter() - t0, 3),
        "generated": datetime.now(timezone.utc).astimezone().isoformat(timespec="seconds"),
        "options": {k: v for k, v in asdict(opt).items() if k not in ("key", "compare", "ignore")},
    })
    return result


def is_identical(result: dict[str, Any]) -> bool:
    c = result["counts"]
    return c["changed"] == 0 and c["added"] == 0 and c["removed"] == 0


# ----------------------------------------------------------------------------
# Column resolution (shared)
# ----------------------------------------------------------------------------

def _resolve_columns(a_cols: list[str], b_cols: list[str], opt: Options):
    missing = [k for k in opt.key if k not in a_cols or k not in b_cols]
    if missing:
        raise CompareError(f"Key column(s) missing from one of the files: {', '.join(missing)}")
    common = [c for c in a_cols if c in b_cols]
    if opt.compare is None:
        compared = [c for c in common if c not in opt.key]
    else:
        bad = [c for c in opt.compare if c not in common]
        if bad:
            raise CompareError(f"Compare column(s) not present in both files: {', '.join(bad)}")
        compared = [c for c in opt.compare if c not in opt.key]
    compared = [c for c in compared if c not in set(opt.ignore)]
    only_a = [c for c in a_cols if c not in b_cols]
    only_b = [c for c in b_cols if c not in a_cols]
    return compared, only_a, only_b


# ----------------------------------------------------------------------------
# DuckDB engine
# ----------------------------------------------------------------------------

def _q(name: str) -> str:
    return '"' + name.replace('"', '""') + '"'


def _lit(s: str) -> str:
    return "'" + s.replace("'", "''") + "'"


def _norm_sql(expr: str, opt: Options) -> str:
    if opt.trim:
        expr = f"trim({expr})"
    if opt.ignore_case:
        expr = f"lower({expr})"
    if opt.empty_is_null:
        expr = f"nullif({expr}, '')"
    return expr


def _diff_sql(a: str, b: str, opt: Options) -> str:
    if opt.tolerance > 0:
        return (f"(CASE WHEN try_cast({a} AS DOUBLE) IS NOT NULL AND try_cast({b} AS DOUBLE) IS NOT NULL "
                f"THEN abs(try_cast({a} AS DOUBLE) - try_cast({b} AS DOUBLE)) > {opt.tolerance!r} "
                f"ELSE ({a} IS DISTINCT FROM {b}) END)")
    return f"({a} IS DISTINCT FROM {b})"


def _compare_duckdb(a_path: str, b_path: str, opt: Options) -> dict[str, Any]:
    import duckdb

    con = duckdb.connect()
    if opt.threads:
        con.execute(f"SET threads = {int(opt.threads)}")
    if opt.memory_limit:
        con.execute(f"SET memory_limit = '{opt.memory_limit}'")
    con.execute("SET preserve_insertion_order = true")

    # null_padding leaves a short row's missing fields absent, which is what the other
    # engines do with them. Without it DuckDB abandons the split on a ragged file and
    # returns it as one column named after the header line, and the comparison then fails
    # with "key column missing" -- a true statement about DuckDB's table and a misleading
    # one about the file.
    read_opts = "all_varchar=true, header=true, sample_size=-1, null_padding=true"
    if opt.delimiter:
        read_opts += f", delim={_lit(opt.delimiter)}"
    if opt.encoding and opt.encoding.lower() not in ("utf-8", "utf8"):
        read_opts += f", encoding={_lit(opt.encoding)}"

    def load(tbl: str, path: str) -> list[str]:
        con.execute(f"CREATE TABLE {tbl}_raw AS SELECT * FROM read_csv({_lit(path)}, {read_opts})")
        return [r[0] for r in con.execute(f"DESCRIBE {tbl}_raw").fetchall()]

    a_cols = load("a", a_path)
    b_cols = load("b", b_path)
    compared, only_a, only_b = _resolve_columns(a_cols, b_cols, opt)
    key = opt.key

    # Normalised projection + stable row number
    for tbl in ("a", "b"):
        proj = ", ".join(f"{_norm_sql(_q(c), opt)} AS {_q(c)}" for c in key + compared)
        con.execute(f"CREATE TABLE {tbl} AS SELECT {proj}, row_number() OVER () AS _rn FROM {tbl}_raw")
        con.execute(f"DROP TABLE {tbl}_raw")

    kq = ", ".join(_q(k) for k in key)
    counts: dict[str, int] = {}
    counts["a_rows"] = con.execute("SELECT count(*) FROM a").fetchone()[0]
    counts["b_rows"] = con.execute("SELECT count(*) FROM b").fetchone()[0]

    # Duplicate keys
    dups = {}
    for tbl in ("a", "b"):
        con.execute(f"CREATE TABLE {tbl}_dup AS SELECT {kq}, count(*) AS n FROM {tbl} GROUP BY ALL HAVING count(*) > 1")
        nk, nr = con.execute(f"SELECT count(*), coalesce(sum(n), 0) FROM {tbl}_dup").fetchone()
        counts[f"{tbl}_dup_keys"], counts[f"{tbl}_dup_rows"] = int(nk), int(nr)
        counts[f"{tbl}_keys"] = counts[f"{tbl}_rows"] - counts[f"{tbl}_dup_rows"] + counts[f"{tbl}_dup_keys"]
        rows = con.execute(f"SELECT * FROM {tbl}_dup ORDER BY n DESC, {kq} LIMIT {opt.max_rows}").fetchall()
        dups[tbl] = {"cols": key + ["count"], "rows": [list(r) for r in rows],
                     "truncated": nk > opt.max_rows}
        # First occurrence per key participates in the join
        con.execute(f"CREATE TABLE {tbl}1 AS SELECT * EXCLUDE(_k) FROM "
                    f"(SELECT *, row_number() OVER (PARTITION BY {kq} ORDER BY _rn) AS _k FROM {tbl}) WHERE _k = 1")

    # Full outer join
    key_sel = ", ".join(f"coalesce(a.{_q(k)}, b.{_q(k)}) AS {_q(k)}" for k in key)
    on = " AND ".join(f"a.{_q(k)} IS NOT DISTINCT FROM b.{_q(k)}" for k in key)
    col_sel, diff_flags = [], []
    for i, c in enumerate(compared):
        a, b = f"a.{_q(c)}", f"b.{_q(c)}"
        d = _diff_sql(a, b, opt)
        col_sel += [f"{a} AS {_q('a_' + str(i))}", f"{b} AS {_q('b_' + str(i))}", f"{d} AS {_q('d_' + str(i))}"]
        diff_flags.append(d)
    changed_expr = " OR ".join(diff_flags) if diff_flags else "false"
    con.execute(f"""
        CREATE TABLE j AS
        SELECT {key_sel},
               CASE WHEN a._rn IS NULL THEN 'added' WHEN b._rn IS NULL THEN 'removed' ELSE 'matched' END AS _status,
               {", ".join(col_sel) + "," if col_sel else ""}
               ({changed_expr}) AS _changed
        FROM a1 a FULL OUTER JOIN b1 b ON {on}
    """)

    for status, chg, n in con.execute("SELECT _status, _changed, count(*) FROM j GROUP BY ALL").fetchall():
        if status == "matched":
            counts["changed" if chg else "unchanged"] = counts.get("changed" if chg else "unchanged", 0) + n
        else:
            counts[status] = counts.get(status, 0) + n
    for k in ("matched", "unchanged", "changed", "added", "removed"):
        counts.setdefault(k, 0)
    counts["matched"] = counts["unchanged"] + counts["changed"]

    # Per-column stats
    columns = []
    if compared:
        agg = ", ".join(
            f"sum({_q('d_%d' % i)}::INT), "
            f"sum(({_q('a_%d' % i)} IS NOT NULL AND {_q('b_%d' % i)} IS NULL)::INT), "
            f"sum(({_q('a_%d' % i)} IS NULL AND {_q('b_%d' % i)} IS NOT NULL)::INT)"
            for i in range(len(compared)))
        stats = con.execute(f"SELECT {agg} FROM j WHERE _status = 'matched'").fetchone()
        for i, c in enumerate(compared):
            columns.append({"name": c, "changed": int(stats[3 * i] or 0),
                            "blanked": int(stats[3 * i + 1] or 0), "filled": int(stats[3 * i + 2] or 0)})

    # Changed rows -> sparse cell diffs
    nk = len(key)
    changed_rows = []
    if compared:
        cols = kq + ", " + ", ".join(f"{_q('a_%d' % i)}, {_q('b_%d' % i)}, {_q('d_%d' % i)}" for i in range(len(compared)))
        cur = con.execute(f"SELECT {cols} FROM j WHERE _status = 'matched' AND _changed ORDER BY {kq} LIMIT {opt.max_rows}")
        for r in cur.fetchall():
            cells = [[i, r[nk + 3 * i], r[nk + 3 * i + 1]] for i in range(len(compared)) if r[nk + 3 * i + 2]]
            changed_rows.append(list(r[:nk]) + [cells])

    def side_rows(status: str, prefix: str):
        cols = kq + "".join(f", {_q(prefix + str(i))}" for i in range(len(compared)))
        rows = con.execute(f"SELECT {cols} FROM j WHERE _status = {_lit(status)} ORDER BY {kq} LIMIT {opt.max_rows}").fetchall()
        return {"cols": key + compared, "rows": [list(r) for r in rows], "truncated": counts[status] > opt.max_rows}

    added = side_rows("added", "b_")
    removed = side_rows("removed", "a_")

    if opt.export_dir:
        os.makedirs(opt.export_dir, exist_ok=True)
        aliases = lambda p: "".join(f", {_q(p + str(i))} AS {_q(c)}" for i, c in enumerate(compared))  # noqa: E731
        con.execute(f"COPY (SELECT {kq}{aliases('b_')} FROM j WHERE _status='added' ORDER BY {kq}) TO "
                    f"{_lit(os.path.join(opt.export_dir, 'added.csv'))} (HEADER)")
        con.execute(f"COPY (SELECT {kq}{aliases('a_')} FROM j WHERE _status='removed' ORDER BY {kq}) TO "
                    f"{_lit(os.path.join(opt.export_dir, 'removed.csv'))} (HEADER)")
        if compared:
            both = "".join(f", {_q('a_%d' % i)} AS {_q(c + ' (A)')}, {_q('b_%d' % i)} AS {_q(c + ' (B)')}"
                           for i, c in enumerate(compared))
            con.execute(f"COPY (SELECT {kq}{both} FROM j WHERE _status='matched' AND _changed ORDER BY {kq}) TO "
                        f"{_lit(os.path.join(opt.export_dir, 'changed.csv'))} (HEADER)")

    con.close()
    return {
        "meta": {"key": key, "compared": compared, "only_in_a": only_a, "only_in_b": only_b,
                 "a_cols": len(a_cols), "b_cols": len(b_cols)},
        "counts": counts,
        "columns": columns,
        "changed": {"cols": key, "rows": changed_rows, "truncated": counts["changed"] > opt.max_rows},
        "added": added, "removed": removed,
        "dup_a": dups["a"], "dup_b": dups["b"],
    }

File: ts/src/engine.ts
/**
 * Composite-key CSV comparison.
 *
 * Engine: DuckDB (default, via @duckdb/node-api). It streams both files from
 * disk, infers delimiters, hash-joins on the composite key and spills to disk
 * when data exceeds RAM, so multi-GB files are fine. The `native` engine is a
 * dependency-free in-memory fallback with the same result contract.
 *
 * The result contract is documented in `types.ts`; both engines must return
 * identical `counts` and `columns` for the same input (CI asserts this).
 */
import { statSync } from "node:fs";
import { basename, resolve } from "node:path";
import { performance } from "node:perf_hooks";

import type { CompareResult, EngineName, EngineResult, Options, Val } from "./types.ts";

export class CompareError extends Error {
  override name = "CompareError";
}

export async function compare(aPath: string, bPath: string, opt: Options): Promise<CompareResult> {
  if (!opt.key || opt.key.length === 0) {
    throw new CompareError("At least one key column is required.");
  }
  for (const p of [aPath, bPath]) {
    if (!isFile(p)) throw new CompareError(`File not found: ${p}`);
  }

  const t0 = performance.now();
  const engine = await resolveEngine(opt.engine);
  let result: EngineResult;
  if (engine === "duckdb") {
    const { compareDuckdb } = await import("./engine-duckdb.ts");
    result = await compareDuckdb(aPath, bPath, opt);
  } else {
    const { compareNative } = await import("./engine-native.ts");
    result = await compareNative(aPath, bPath, opt);
  }

  const { key: _k, compare: _c, ignore: _i, ...options } = opt;
  return {
    ...result,
    meta: {
      ...result.meta,
      a: fileMeta(aPath),
      b: fileMeta(bPath),
      engine,
      seconds: Math.round(performance.now() - t0) / 1000,
      generated: isoWithOffset(new Date()),
      options,
    },
  };
}

export function isIdentical(result: CompareResult): boolean {
  const c = result.counts;
  return c.changed === 0 && c.added === 0 && c.removed === 0;
}

export async function resolveEngine(engine: EngineName): Promise<"duckdb" | "native"> {
  if (engine === "duckdb" || engine === "native") return engine;
  if (engine !== "auto") throw new CompareError(`Unknown engine: ${engine as string}`);
  try {
    await import("@duckdb/node-api");
    return "duckdb";
  } catch {
    return "native";
  }
}

// ----------------------------------------------------------------------------
// Shared helpers (used by both engines)
// ----------------------------------------------------------------------------

export interface ResolvedColumns {
  compared: string[];
  onlyA: string[];
  onlyB: string[];
}

export function resolveColumns(aCols: string[], bCols: string[], opt: Options): ResolvedColumns {
  const aSet = new Set(aCols);
  const bSet = new Set(bCols);
  const missing = opt.key.filter((k) => !aSet.has(k) || !bSet.has(k));
  if (missing.length) {
    throw new CompareError(`Key column(s) missing from one of the files: ${missing.join(", ")}`);
  }
  const common = aCols.filter((c) => bSet.has(c));
  const commonSet = new Set(common);
  const keySet = new Set(opt.key);
  let compared: string[];
  if (opt.compare === null) {
    compared = common.filter((c) => !keySet.has(c));
  } else {
    const bad = opt.compare.filter((c) => !commonSet.has(c));
    if (bad.length) {
      throw new CompareError(`Compare column(s) not present in both files: ${bad.join(", ")}`);
    }
    compared = opt.compare.filter((c) => !keySet.has(c));
  }
  const ignore = new Set(opt.ignore);
  compared = compared.filter((c) => !ignore.has(c));
  return {
    compared,
    onlyA: aCols.filter((c) => !bSet.has(c)),
    onlyB: bCols.filter((c) => !aSet.has(c)),
  };
}

/** Apply --trim / --ignore-case / --empty-is-null to one value (native engine). */
export function normValue(v: Val, opt: Options): Val {
  if (v === null) return null;
  let s = v;
  if (opt.trim) s = s.trim();
  if (opt.ignore_case) s = s.toLowerCase();
  if (opt.empty_is_null && s === "") return null;
  return s;
}

const NUMBER = /^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/;

function asNumber(v: Val): number | null {
  if (v === null) return null;
  const s = v.trim();
  return NUMBER.test(s) ? Number(s) : null;
}

/** Cell inequality with optional numeric tolerance (native engine). */
export function differs(x: Val, y: Val, opt: Options): boolean {
  if (x === null && y === null) return false;
  if (opt.tolerance > 0) {
    const nx = asNumber(x);
    const ny = asNumber(y);
    if (nx !== null && ny !== null) return Math.abs(nx - ny) > opt.tolerance;
  }
  return x !== y;
}

/** Sort key values the way DuckDB orders VARCHAR: ascending, NULLs last. */
export function compareKeys(a: Val[], b: Val[], nk: number): number {
  for (let i = 0; i < nk; i++) {
    const x = a[i] ?? null;
    const y = b[i] ?? null;
    if (x === y) continue;
    if (x === null) return 1;
    if (y === null) return -1;
    return x < y ? -1 : 1;
  }
  return 0;
}

function isFile(p: string): boolean {
  try {
    return statSync(p).isFile();
  } catch {
    return false;
  }
}

function fileMeta(p: string) {
  return { name: basename(p), path: resolve(p), size: statSync(p).size };
}

/** Local time with numeric offset, seconds precision: 2026-09-03T18:56:28+00:00 */
export function isoWithOffset(d: Date): string {
  const pad = (n: number) => String(n).padStart(2, "0");
  const off = -d.getTimezoneOffset();
  const sign = off >= 0 ? "+" : "-";
  const abs = Math.abs(off);
  return (
    `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}T` +
    `${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}` +
    `${sign}${pad(Math.floor(abs / 60))}:${pad(abs % 60)}`
  );
}
